### PR TITLE
[Feature] Azure Maps Account Support

### PR DESCRIPTION
**Pull request by Krueladin**
Adding Azure Map Accounts support to Terraform.

Adds azurerm_maps_account data source.
Adds azurerm_maps_account resource type.
Adds website documentation for data source and resource.
Adds data source and resource acceptance tests.

### CODE REVIEW

**Comment by Krueladin** [comment]
WIP

**Comment by Krueladin** [comment]
Ready for review

File: azurerm/data_source_maps_account.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+func dataSourceArmMapsAccount() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceMapsAccountRead,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validate.ApiManagementApiName,
```
Comment (katbyte) [inline azurerm/data_source_maps_account.go]: Could we use a validation function specifically for this property?

File: azurerm/data_source_maps_account.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+		return fmt.Errorf("Error making Read request on Maps Account %q (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
+	d.SetId(*resp.ID)
+
+	d.Set("name", name)
+	d.Set("resource_group_name", resourceGroup)
+	if sku := resp.Sku; sku != nil {
+		d.Set("sku", string(*sku.Name))
```
Comment (katbyte) [inline azurerm/data_source_maps_account.go]: could we change this to:
```suggestion
		d.Set("sku_name", sku.Name)
```

File: azurerm/data_source_maps_account.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validate.ApiManagementApiName,
+			},
+
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"sku": {
```
Comment (katbyte) [inline azurerm/data_source_maps_account.go]: Given we are just using the sku.name here could we rename this to `sku_name`?

File: azurerm/resource_arm_maps_account.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"resource_group_name": azure.SchemaResourceGroupName(),
+
+			"sku": {
```
Comment (katbyte) [inline azurerm/resource_arm_maps_account.go]: As above could we change this to `sku_name`

File: website/docs/r/maps_account.html.markdown [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+resource "azurerm_resource_group" "test" {
+  name     = "example-resources"
+  location = "West Europe"
+}
+
+resource "azurerm_maps_account" "test" {
+  name                = "example-maps-account"
+  resource_group_name = "${azurerm_resource_group.test.name}"
+
+  sku = "S1"
```
Comment (katbyte) [inline website/docs/r/maps_account.html.markdown]:
```suggestion
  sku_name = "S1"
```

File: website/docs/r/maps_account.html.markdown [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the Azure Maps Account. Changing this forces a new resource to be created.
+
+* `resource_group_name` - (Required) The name of the Resource Group in which the Azure Maps Account should exist. Changing this forces a new resource to be created.
+
+* `sku` - (Optional) The sku of the Azure Maps Account. Values available are S0, S1. Defaults to S0.
```
Comment (katbyte) [inline website/docs/r/maps_account.html.markdown]:
```suggestion
* `sku_name` - (Optional) The sku of the Azure Maps Account. Values available are `S0`, `S1`. Defaults to `S0`.
```

File: website/docs/d/maps_account.html.markdown [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+* `x_ms_client_id` - A unique identifier for the Maps Account.
+
+* `primary_access_key` - The primary key used to authenticate and authorize access to the Maps REST APIs.
+
+* `secondary_access_key` - The primary key used to authenticate and authorize access to the Maps REST APIs. The second key is given to provide seemless key regeneration.
+
+* `x_ms_client_id` - A unique identifier for the Maps Account.
+
+* `sku` - The sku of the Azure Maps Account.
```
Comment (katbyte) [inline website/docs/d/maps_account.html.markdown]: Could we move this up to under `id`

File: go.mod [hunk 8]
```diff
@@ -1,7 +1,7 @@
 module github.com/terraform-providers/terraform-provider-azurerm
 
 require (
-	github.com/Azure/azure-sdk-for-go v30.0.0+incompatible
+	github.com/Azure/azure-sdk-for-go v30.1.0+incompatible
```
Comment (katbyte) [inline go.mod]: this is an SDK upgrade, we do these as separate PRs. Could we revert this change?

File: azurerm/resource_arm_maps_account_test.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+		if resp.StatusCode != http.StatusNotFound {
+			return fmt.Errorf("Maps Account still exists:\n%#v", resp.ID)
+		}
+	}
+
+	return nil
+
+}
+
+func accountNamePrefix() string {
```
Comment (katbyte) [inline azurerm/resource_arm_maps_account_test.go]: COuld we just hard code this instead?

File: azurerm/resource_arm_maps_account.go [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+	name := id.Path["accounts"]
+
+	if _, err := client.Delete(ctx, resGroup, name); err != nil {
+		return fmt.Errorf("Error deleting Maps Account %q (Resource Group %q): %+v", name, resGroup, err)
+	}
+
+	return nil
+}
+
+func globalLocation() string {
```
Comment (katbyte) [inline azurerm/resource_arm_maps_account.go]: Could we hard code this instead of having a function?

File: azurerm/resource_arm_maps_account.go [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+			return nil
+		}
+
+		return fmt.Errorf("Error making Read request on Maps Account %q (Resource Group %q): %+v", name, resGroup, err)
+	}
+
+	d.Set("name", resp.Name)
+	d.Set("resource_group_name", resGroup)
+	if sku := resp.Sku; sku != nil {
+		d.Set("sku", string(*sku.Name))
```
Comment (katbyte) [inline azurerm/resource_arm_maps_account.go]: d.Set deals with nil values so could we do this instead:
```suggestion
		d.Set("sku_name", sku.Name)
```

File: azurerm/resource_arm_maps_account.go (hunk 10 above)
Comment (Krueladin) [reply azurerm/resource_arm_maps_account.go]: Is there anywhere the Terraform provider keeps constants like this?

File: azurerm/data_source_maps_account.go (hunk 1 above)
Comment (Krueladin) [reply azurerm/data_source_maps_account.go]: Looks like I didn't catch this here. I use a lot of existing code to help drive new additions. Will update.

File: website/docs/d/maps_account.html.markdown [hunk 12]
```diff
@@ -0,0 +1,43 @@
+---
```
Comment (tombuildsstuff) [inline website/docs/d/maps_account.html.markdown]: can we also add this to the sidebar? https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/website/azurerm.erb#L119

**Comment by Krueladin** [comment]
@tombuildsstuff Go ahead. I haven't had time to make the requested changes and looks like you've got the momentum here :)

**Comment by katbyte** [comment]
Seems like microsoft is returning the `sku_names` in lowercase 😞 
```
------- Stdout: -------
=== RUN   TestAccAzureRMMapsAccount_basic
=== PAUSE TestAccAzureRMMapsAccount_basic
=== CONT  TestAccAzureRMMapsAccount_basic
--- FAIL: TestAccAzureRMMapsAccount_basic (77.08s)
    testing.go:568: Step 0 error: Check failed: Check 5/5 error: azurerm_maps_account.test: Attribute 'sku_name' expected "S0", got "s0"
    testing.go:629: Error destroying resource! WARNING: Dangling resources
        may exist. The full state and error is shown below.
        
        Error: Check failed: Maps Account still exists:
        (*string)(nil)
        
        State: <no state>
FAIL
```

**Comment by tombuildsstuff** [comment]
Updated the casing on the SKU - tests pass:

```
$ acctests azurerm TestAccAzureRMMapsAccount_
=== RUN   TestAccAzureRMMapsAccount_basic
=== PAUSE TestAccAzureRMMapsAccount_basic
=== RUN   TestAccAzureRMMapsAccount_sku
=== PAUSE TestAccAzureRMMapsAccount_sku
=== RUN   TestAccAzureRMMapsAccount_tags
=== PAUSE TestAccAzureRMMapsAccount_tags
=== CONT  TestAccAzureRMMapsAccount_basic
=== CONT  TestAccAzureRMMapsAccount_tags
--- PASS: TestAccAzureRMMapsAccount_basic (80.43s)
--- PASS: TestAccAzureRMMapsAccount_tags (111.23s)
=== CONT  TestAccAzureRMMapsAccount_sku
--- PASS: TestAccAzureRMMapsAccount_sku (82.26s)
PASS
ok  	github.com/terraform-providers/terraform-provider-azurerm/azurerm	273.956s
```
